examples/player/simid_player.js

```diff
@@ -609,16 +609,75 @@ class SimidPlayer {
     console.log('The creative has asked for the player to ping ' + requestedUrlArray);
   }
 
+  /**
+   * Called when video playback is complete.
+   * @private
+   */
+  videoComplete() {
+    this.simidProtocol.sendMessage(MediaMessage.ENDED);
+
+    //If the request duration is longer than the ad duration, the ad extends for the requested amount of time
+    if (this.requestedDuration_ != UNLIMITED_DURATION) {
+      const durationChangeMs = (this.requestedDuration_ - this.adVideoElement_.duration) * 1000;
+      setTimeout(() => {
+        this.stopAd(StopCode.CREATIVE_INITIATED);
+      }, durationChangeMs);
+    }
+
+    if (this.requestedDuration_ == NO_REQUESTED_DURATION) {
+      this.stopAd(StopCode.MEDIA_PLAYBACK_COMPLETE);
+    }
+  }
+
+  /**
+   * Called when creative requests a change in duration of ad.
+   * @private
+   */
   onRequestChangeAdDuration(incomingMessage) {
-    if (this.requestedDuration_ != NO_REQUESTED_DURATION) {
-      // TODO: Support multiple change duration requests.
-      this.simidProtocol.reject(incomingMessage);
+    const originalDuration = this.requestedDuration_;
+    this.requestedDuration_  = incomingMessage.args['duration'];
+    if (this.requestedDuration_ != UNLIMITED_DURATION && this.requestedDuration_< 0) {
+      const durationErrorMessage = {
+        errorCode: PlayerErrorCode.UNSUPPORTED_TIME,
+        message: 'A negative duration is not valid.'
+      }
+      this.requestedDuration_ = originalDuration;
+      this.simidProtocol.reject(incomingMessage, durationErrorMessage);
+    } else if (this.requestedDuration_ == UNLIMITED_DURATION) {
+      //Note: Users can end the ad using the close ad button on the player
+      return;
+    } else {
+      //If requested duration is any other acceptable value
+      this.changeAdEndTime();
+      this.simidProtocol.resolve(incomingMessage);
     }
-    const requestedDuration  = incomingMessage.args['duration'];
-    this.requestedDuration_ = requestedDuration;
-    this.simidProtocol.resolve(incomingMessage);
   }
 
+  /**
+   * Compares the requested change duration with the current ad time
+   *  to determine when to stop the ad, every 250 milliseconds.
+   * @private
+   */
+  changeAdEndTime() {
+    this.durationInterval_ = setInterval(() => {
+        this.compareAdAndRequestedDurations(); 
+    }, INTERVAL_TIMER);
+  }
+
+  /**
+   * Compares the duration of the ad with the requested change duration.
+   * If request duration is shorter, the ad stops early. 
+   * @private
+   */
+  compareAdAndRequestedDurations() {
+    if (this.adVideoElement_.currentTime >= this.requestedDuration_) {
```

Comment:
You might hate me, but I promise we're working towards better code. What do you think of just putting that into the currentTimeUpdate callback?

We wouldn't need this interval at all anymore.  You would have to add one more conditional there that we weren't NO_REQUESTED_DURATION or UNLIMITED_DURATION.  But you could probably delete a bunch of complexity and keep this code more simple.


Basically, you would modify the code on line 563 (of the current player). Look for this:
    this.videoTrackingEvents_.set("timeupdate", () => {

You would change it to this:
this.videoTrackingEvents_.set("timeupdate", () => {
      this.simidProtocol.sendMessage(MediaMessage.TIME_UPDATE,
        {'currentTime': this.adVideoElement_.currentTime});
       this.compareAdAndRequestedDuration();
    });

Then modify your compareAdAndRequestedDuration function with this:
compareAdAndRequestedDurations() {
   if (this.requestedDuration_ == NO_REQUESTED_DURATION ||
    this.requestedDuration_ == UNLIMITED_DURATION) {
   // We don't terminate the ad in these cases here.
    return;
}
    if (this.adVideoElement_.currentTime >= this.requestedDuration_) {
///.... the same code you have...


Then you would get rid of the line:

this.changeAdEndTime();

and you would get rid of the function
changeAdEndTime

You wouldn't need those anymore, because any time any time update happened you would have a call to compareAdAndRequestedDurations

Reply:
Happy to keep learning better coding practices. Just made the changes, let me know your thoughts.